### PR TITLE
Add sphum limiter and precipitation update to reservoir stepper

### PR DESCRIPTION
**Pull request by AnnaKwa**
This adds the positive specific humidity limiter and a precipitation update to the reservoir stepper. It also adds the associated diagnostics for column net heating/moistening from the reservoir step.

The existing functions for these computations already existed for the ML stepper but took in tendency inputs only. To minimize changes to the existing code I added a conversion of the reservoir's state prediction to the equivalent tendencies, which are now returned by the reservoir stepper instead of an empty dict. The tendencies are used in the reservoir step of the loop file to compute diagnostics.  

Significant internal changes:
- added `hydrostatic` and `mse_conserving_limiter` optional fields in the reservoir config. These are used to compute diagnostics in the same way as in the ML stepper, where they are also config fields.
- added `model_timestep` attribute to the reservoir stepper so it can compute the equivalent tendencies from the predicted and input state.
 
- [x ] Tests added

Resolves #<github issues> [JIRA-TAG]

Coverage reports (updated automatically):

### CODE REVIEW

File: workflows/prognostic_c48_run/runtime/diagnostics/compute.py [hunk 1]
```diff
@@ -24,8 +24,9 @@ def enforce_heating_and_moistening_tendency_constraints(
     timestep: float,
     hydrostatic: bool,
     mse_conserving: bool,
-    temperature_tendency_name="dQ1",
-    humidity_tendency_name="dQ2",
+    temperature_tendency_name: str = "dQ1",
+    humidity_tendency_name: str = "dQ2",
+    zero_fill_missing_tendencies: bool = False,
```
Comment (AnnaKwa) [inline workflows/prognostic_c48_run/runtime/diagnostics/compute.py:29]: This is added as an optional arg to preserve prior behavior in the ML postphysics stepper where these diags aren't output if dQ1, dQ2 are not present. If this isn't important I can remove it and the zero-filled fields will always be in the diags.

File: workflows/prognostic_c48_run/runtime/loop.py [hunk 2]
```diff
@@ -574,12 +578,43 @@ def _increment_reservoir(self) -> Diagnostics:
...
 
     def _apply_reservoir_update_to_state(self) -> Diagnostics:
-        # TODO: handle tendencies
+        # TODO: handle tendencies. Currently the returned tendencies
+        # are only used for diagnostics and are not used in updating state
         if self._reservoir_predict_stepper is not None:
-            [_, diags, state] = self._reservoir_predict_stepper(
-                self._state.time, self._state
+            [
+                tendencies_from_state_prediction,
```
Comment (AnnaKwa) [inline workflows/prognostic_c48_run/runtime/loop.py:585]: The reservoir stepper now returns a non empty tendency dict (they are calculated using the state prediction). These are used to compute diagnostics, but are not used in the fv3 state update.

File: workflows/prognostic_c48_run/runtime/diagnostics/compute.py (hunk 1 above)
Comment (frodre) [reply workflows/prognostic_c48_run/runtime/diagnostics/compute.py:29]: 👍 I think it's best to preserve previous behavior when possible.

File: workflows/prognostic_c48_run/runtime/loop.py [hunk 3]
```diff
@@ -574,12 +578,43 @@ def _increment_reservoir(self) -> Diagnostics:
...
+                diags,
+                state_updates,
+            ] = self._reservoir_predict_stepper(self._state.time, self._state)
+            (
```
Comment (frodre) [inline workflows/prognostic_c48_run/runtime/loop.py:589]: The smallest of small nits,  consistent use of `[` vs `(` to group variable assignments.

File: workflows/prognostic_c48_run/runtime/loop.py [hunk 4]
```diff
@@ -574,12 +578,43 @@ def _increment_reservoir(self) -> Diagnostics:
...
+                stepper_diags,
+                net_moistening,
+            ) = self._reservoir_predict_stepper.get_diagnostics(
+                self._state, tendencies_from_state_prediction
+            )
+            diags.update(stepper_diags)
+            if self._reservoir_predict_stepper.diagnostic is True:  # type: ignore
```
Comment (frodre) [inline workflows/prognostic_c48_run/runtime/loop.py:596]: nit: `if self._reservoir_predict_stepper.diagnostic:`

Comment (frodre) [reply workflows/prognostic_c48_run/runtime/loop.py:596]: Could rename the underlying attribute to `is_diagnostic` to make it clearer that it's a bool

File: workflows/prognostic_c48_run/runtime/loop.py [hunk 5]
```diff
@@ -574,12 +578,43 @@ def _increment_reservoir(self) -> Diagnostics:
...
+                rename_diagnostics(diags, label="reservoir_predictor")
+
+            state_updates[TOTAL_PRECIP] = precipitation_sum(
```
Comment (frodre) [inline workflows/prognostic_c48_run/runtime/loop.py:599]: Just checking that it's okay to always provide `TOTAL_PRECIP` updates.  I'm guessing based on the signature that an update with 0-tendency net moistening returns the same precip from the state.

**Review by frodre — APPROVED** [review]
LGTM, just a couple very minor adjustments and one question.

File: workflows/prognostic_c48_run/runtime/loop.py (hunk 5 above)
Comment (AnnaKwa) [reply workflows/prognostic_c48_run/runtime/loop.py:599]: Correct.

**Pull request merged**